File: backend/api/summary.py
"""Сводка за период: 24ч / Неделя / Месяц, по всем источникам сразу."""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from backend.db import get_db
from backend.models import Message, User
from backend.queries import level_distribution, summary_period_start, user_messages
from backend.schemas import SummaryOut
from backend.security import current_user
from backend.serializers import message_brief

router = APIRouter(prefix="/api", tags=["summary"])

TOP_LIMIT = 4  # «Главное за период» — до 4 сообщений


@router.get("/summary", response_model=SummaryOut)
def get_summary(
    period: str = Query(default="24h", pattern="^(24h|week|month)$"),
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> SummaryOut:
    stmt = user_messages(user).where(Message.received_at >= summary_period_start(period))
    messages = db.execute(
        stmt.order_by(Message.received_at.desc(), Message.id.desc())
    ).scalars().all()

    top = [m for m in messages if m.effective_level == "CRITICAL"]
    top += [m for m in messages if m.effective_level == "HIGH"]

    return SummaryOut(
        period=period,
        total=len(messages),
        distribution=level_distribution(db, stmt),
        needs_reply=sum(1 for m in messages if m.needs_reply),
        needs_action=sum(1 for m in messages if m.needs_action),
        top=[message_brief(m) for m in top[:TOP_LIMIT]],
    )

File: backend/internal/api/summary.go
package api

import (
	"net/http"
	"time"

	"personalinbox/internal/store"
	"personalinbox/internal/view"
)

// topLimit — «Главное за период» показывает до 4 сообщений.
const topLimit = 4

// handleSummary — сводка за период по всем источникам сразу.
func (s *Server) handleSummary(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	period, ok := param(w, r, "period", "24h", []string{"24h", "week", "month"})
	if !ok {
		return
	}

	messages, err := s.db.MessagesSince(user.ID, store.SummaryPeriodStart(period, time.Time{}))
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось собрать сводку")
		return
	}

	needsReply, needsAction := 0, 0
	for _, message := range messages {
		if message.NeedsReply {
			needsReply++
		}
		if message.NeedsAction {
			needsAction++
		}
	}

	// Сначала все CRITICAL, потом HIGH — и не больше четырёх.
	top := make([]view.MessageBrief, 0, topLimit)
	for _, level := range []string{"CRITICAL", "HIGH"} {
		for _, message := range messages {
			if len(top) >= topLimit {
				break
			}
			if message.EffectiveLevel() == level {
				top = append(top, view.MessageBriefOut(message))
			}
		}
	}

	respond(w, http.StatusOK, view.Summary{
		Period:       period,
		Total:        len(messages),
		Distribution: store.Distribution(messages),
		NeedsReply:   needsReply,
		NeedsAction:  needsAction,
		Top:          top,
	})
}
